File: app/src/main/java/com/abaffym/moviesapp/ui/MoviesRepository.java
package com.abaffym.moviesapp.ui;

import com.abaffym.moviesapp.data.local.MoviesLocalDataSource;
import com.abaffym.moviesapp.data.remote.entity.DiscoverMoviesResponse;
import com.abaffym.moviesapp.data.remote.entity.MovieEntity;
import com.abaffym.moviesapp.data.remote.rest.MovieMapper;
import com.abaffym.moviesapp.data.remote.rest.MoviesApi;
import com.abaffym.moviesapp.di.NetworkModule;
import com.abaffym.moviesapp.model.Movie;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;

public class MoviesRepository {

    private MoviesApi moviesApi;

    private MovieMapper movieMapper;

    public MoviesRepository() {
        moviesApi = NetworkModule.INSTANCE.moviesApi();
        movieMapper = new MovieMapper();
    }

    private Single<List<Movie>> getChachedMovies() {
        return Single.fromCallable(new Callable<List<Movie>>() {
            @Override
            public List<Movie> call() {
                return MoviesLocalDataSource.INSTANCE.getCachedMovies();
            }
        });
    }

    private Single<List<Movie>> getRemoteMovies() {
        return moviesApi.discoverMovies()
                .map(new Function<DiscoverMoviesResponse<MovieEntity>, List<Movie>>() {
                    @Override
                    public List<Movie> apply(DiscoverMoviesResponse<MovieEntity> response) {
                        List<Movie> movies = new ArrayList<>();
                        for (MovieEntity entity : response.getResults()) {
                            movies.add(movieMapper.mapEntityToModel(entity));
                        }
                        return movies;
                    }
                }).doOnSuccess(new Consumer<List<Movie>>() {
                    @Override
                    public void accept(List<Movie> movies) {
                        storeInLocalCache(movies);
                    }
                });
    }

    public Single<List<Movie>> getAllMovies() {
        return Single.concat(getChachedMovies(), getRemoteMovies())
                .filter(new Predicate<List<Movie>>() {
                    @Override
                    public boolean test(List<Movie> movies) {
                        return !movies.isEmpty();
                    }
                })
                .firstOrError()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<Movie>> getFavoriteMovies() {
        return Single.fromCallable(new Callable<List<Movie>>() {
            @Override
            public List<Movie> call() {
                return MoviesLocalDataSource.INSTANCE.getFavoriteMovies();
            }
        })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Movie getMovieById(long id) {
        for (Movie movie : MoviesLocalDataSource.INSTANCE.getCachedMovies()) {
            if (movie.getId() == id) {
                return movie;
            }
        }
        throw new IllegalStateException("MovieEntity not found");
    }

    public boolean isFavorite(long movieId) {
        return MoviesLocalDataSource.INSTANCE.isFavorite(movieId);
    }

    public void addToFavorites(long movieId) {
        MoviesLocalDataSource.INSTANCE.addFavoriteMovie(movieId);
    }

    public void removeFromFavorites(Long movieId) {
        MoviesLocalDataSource.INSTANCE.removeFavoriteMovie(movieId);
    }

    private void storeInLocalCache(List<Movie> movies) {
        MoviesLocalDataSource.INSTANCE.setCachedMovies(movies);
    }
}

File: app/src/main/java/com/abaffym/moviesapp/ui/MoviesRepository.kt
package com.abaffym.moviesapp.ui

import com.abaffym.moviesapp.data.local.MoviesLocalDataSource
import com.abaffym.moviesapp.data.remote.rest.MovieMapper
import com.abaffym.moviesapp.di.NetworkModule
import com.abaffym.moviesapp.model.Movie
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class MoviesRepository {

    private val moviesApi = NetworkModule.moviesApi()

    private val movieMapper = MovieMapper()

    private val cachedMovies: Single<List<Movie>>
        get() = Single.fromCallable { MoviesLocalDataSource.cachedMovies }

    private val remoteMovies: Single<List<Movie>>
        get() = moviesApi.discoverMovies()
                .map { response -> response.results.map(movieMapper::mapEntityToModel) }
                .doOnSuccess { movies -> storeInLocalCache(movies) }

    val allMovies: Single<List<Movie>>
        get() = Single.concat(cachedMovies, remoteMovies)
                .filter { movies -> !movies.isEmpty() }
                .firstOrError()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())

    val favoriteMovies: Single<List<Movie>>
        get() = Single.fromCallable { MoviesLocalDataSource.favoriteMovies }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())

    fun getMovieById(id: Long) = MoviesLocalDataSource.cachedMovies
            .firstOrNull { movie -> movie.id == id }
            ?: throw IllegalStateException("MovieEntity not found")


    fun isFavorite(movieId: Long) = MoviesLocalDataSource.isFavorite(movieId)

    fun addToFavorites(movieId: Long) = MoviesLocalDataSource.addFavoriteMovie(movieId)

    fun removeFromFavorites(movieId: Long?) = MoviesLocalDataSource.removeFavoriteMovie(movieId!!)

    private fun storeInLocalCache(movies: List<Movie>) = MoviesLocalDataSource.setCachedMovies(movies)
}
